api/rest/bundle_handler.go

```diff
@@ -0,0 +1,354 @@
+
+	//TODO(janisz): Think about removing/replacing summary files with something more structured
+	summaryErrorsReportFileName = "summaryErrorsReport.txt" // error log in bundle
+	summaryReportFileName       = "summaryReport.txt"       // error log in bundle
+
+	filePerm = 0600
+	dirPerm  = 0700
+)
+
+type Bundle struct {
+	ID      string    `json:"id,omitempty"`
+	Size    int64     `json:"size,omitempty"` // length in bytes for regular files; 0 when Canceled or Deleted
+	Status  Status    `json:"status"`
+	Started time.Time `json:"started_at,omitempty"`
+	Stopped time.Time `json:"stopped_at,omitempty"`
+	Errors  []string  `json:"errors,omitempty"`
+}
+
+type ErrorResponse struct {
+	Code  int    `json:"code"`
+	Error string `json:"error"`
+}
+
+type Clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (realClock) Now() time.Time { return time.Now() }
+
+func NewBundleHandler(workDir string, collectors []collectors.Collector) BundleHandler {
+	return BundleHandler{
+		clock:      realClock{},
+		workDir:    workDir,
+		collectors: collectors,
+	}
+}
+
+// BundleHandler is a struct that collects all functions
+// responsible for diagnostics bundle lifecycle
+type BundleHandler struct {
+	clock      Clock
+	workDir    string                 // location where bundles are generated and stored
+	collectors []collectors.Collector // information what should be in the bundle
+}
+
+func (h BundleHandler) Create(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	_, err := h.getBundleState(id)
+	if err == nil {
+		writeJSONError(w, http.StatusConflict, fmt.Errorf("bundle %s already exists", id))
+		return
+	}
+
+	bundleWorkDir := filepath.Join(h.workDir, id)
+	err = os.MkdirAll(bundleWorkDir, dirPerm)
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not Create bundle %s workdir: %s", id, err))
+		return
+	}
+
+	stateFilePath := filepath.Join(h.workDir, id, stateFileName)
+
+	bundle := Bundle{
+		ID:      id,
+		Started: h.clock.Now(),
+		Status:  Started,
+	}
+
+	bundleStatus := jsonMarshal(bundle)
+	err = ioutil.WriteFile(stateFilePath, bundleStatus, filePerm)
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not update state file %s: %s", id, err))
+		return
+	}
+
+	dataFile, err := os.Create(filepath.Join(h.workDir, id, dataFileName))
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not create data file %s: %s", id, err))
+	}
+
+	ctx := context.Background() //TODO(janisz): Use context with deadline
+	done := make(chan []string)
+
+	go collectAll(ctx, done, dataFile, h.collectors)
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			break
+		case bundle.Errors = <-done:
+			bundle.Status = Done
+			bundle.Stopped = h.clock.Now()
+			newBundleStatus := jsonMarshal(bundle)
+			e := ioutil.WriteFile(stateFilePath, newBundleStatus, filePerm)
```

Comment:
you can just inline the `jsonMarshal(bundle)` call here and get rid of `newBundleStatus`.